### PR TITLE
Add benchmarks

### PR DESCRIPTION
**Pull request by flying-sheep**
- [ ] Closes #
- [x] Tests included or not required because:

- [ ] Release notes not necessary because:

Builds on and supersedes #2482

### CODE REVIEW

**Comment by flying-sheep** [comment]
pre-commit.ci autofix

File: benchmarks/asv.conf.json [hunk 1]
```diff
@@ -81,7 +81,7 @@
         "zarr": [""],
         "pytest": [""],
         "scanpy": [""],
-        "igraph": [""],
+        "igraph-python": [""],
```
Comment (ivirshup) [inline benchmarks/asv.conf.json]: Below, this is written `python-igraph`. Is that important?

Comment (flying-sheep) [reply benchmarks/asv.conf.json]: turns out yes, you seem to have commented on an old commit where I haven’t fixed that yet.

`python-igraph` is correct for conda. `python-igraph` also used to be the old PyPI name which is now deprecated.

File: benchmarks/benchmarks/readwrite.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+* Read time, write time
+* Peak memory during io
+* File sizes
+
+Parameterized by:
+
+* What method is being used
+* What data is being included
+* Size of data being used
```
Comment (ilan-gold) [inline benchmarks/benchmarks/readwrite.py]: Are any of these parametrized?

Comment (ilan-gold) [reply benchmarks/benchmarks/readwrite.py]: Or is this just a wish-list?

File: benchmarks/benchmarks/readwrite.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    def track_read_full_memratio(self, *_):
+        mem_recording = memory_usage(
+            (sedate(anndata.read_h5ad, seconds=0.005), (self.path,)), interval=0.001
+        )
+
+        base_size = mem_recording[-1] - mem_recording[0]
+        print(np.max(mem_recording) - np.min(mem_recording))
+        print(base_size)
+        return (np.max(mem_recording) - np.min(mem_recording)) / base_size
```
Comment (ilan-gold) [inline benchmarks/benchmarks/readwrite.py]: Why a print here?

File: benchmarks/benchmarks/tools.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        sc.tl.diffmap(self.adata)
+
+    def peakmem_diffmap(self, *_):
+        sc.tl.diffmap(self.adata)
+
+    # def time_embedding_density(self, *_):
+    #     sc.tl.embedding_density(self.adata, groupby="batch")
+
+    # def peakmem_embedding_density(self, *_):
+    #     sc.tl.embedding_density(self.adata, groupby="batch")
```
Comment (ilan-gold) [inline benchmarks/benchmarks/tools.py]: ?

File: benchmarks/benchmarks/tools.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    #     sc.tl.paga(self.adata, groups='leiden')
+
+    # def peakmem_paga(self, *_):
+    #     sc.tl.paga(self.adata, groups='leiden')
+
+    # def time_draw_graph(self, *_):
+    #     sc.tl.draw_graph(self.adata)
+
+    # def peakmem_draw_graph(self, *_):
+    #     sc.tl.draw_graph(self.adata)
```
Comment (ilan-gold) [inline benchmarks/benchmarks/tools.py]: ?

File: benchmarks/benchmarks/readwrite.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        adata = anndata.read_h5ad(self.path)
+        adata_size = sys.getsizeof(adata)
+
+        return adata_size
+
+    def track_actual_in_memory_size(self, *_):
+        adata = anndata.read_h5ad(self.path)
+        adata_size = get_actualsize(adata)
+
+        return adata_size
```
Comment (ilan-gold) [inline benchmarks/benchmarks/readwrite.py]: Why are we benchmarking anndata i/o here?

File: benchmarks/benchmarks/utils.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        for obj in objects:
+            if id(obj) not in ids:
+                ids.add(id(obj))
+                memory_size += sys.getsizeof(obj)
+                new.append(obj)
+        objects = gc.get_referents(*new)
+    return memory_size
+
+
+def get_anndata_memsize(adata: AnnData) -> float:
```
Comment (ilan-gold) [inline benchmarks/benchmarks/utils.py]: Is this function used anywhere?

File: benchmarks/benchmarks/utils.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
+from __future__ import annotations
```
Comment (ilan-gold) [inline benchmarks/benchmarks/utils.py:1]: Is this file copy-pasted (mostly) from the anndata repo?  A lot of these functions don't appear to be used at all.

Comment (ilan-gold) [reply benchmarks/benchmarks/utils.py:1]: For example, I don't see anywhere `take_view` is used outside of its declaration in this file.

**Review by ilan-gold — DISMISSED** [review]

File: benchmarks/benchmarks/readwrite.py (hunk 2 above)
Comment (flying-sheep) [reply benchmarks/benchmarks/readwrite.py]: we just have the dataset “parametrized”, and we only specify one. I feel like that is more like an example of how parametrization looks like too. I could just simplify it to not use a parameter.

File: benchmarks/benchmarks/readwrite.py (hunk 3 above)
Comment (flying-sheep) [reply benchmarks/benchmarks/readwrite.py]: good point, removed

File: benchmarks/benchmarks/readwrite.py (hunk 6 above)
Comment (flying-sheep) [reply benchmarks/benchmarks/readwrite.py]: great point. seems like this stuff has just been copied over

File: benchmarks/benchmarks/utils.py (hunk 7 above)
Comment (flying-sheep) [reply benchmarks/benchmarks/utils.py]: no, I’ll remove it.

**Pull request merged**